sdk/messaging/azservicebus/sender.go

```diff
@@ -5,36 +5,36 @@ package azservicebus
 
 import (
 	"context"
-	"sync"
+	"fmt"
 
 	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus/internal"
 	"github.com/Azure/go-amqp"
 	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/devigned/tab"
 )
 
-// SenderOption specifies an option that can configure a Sender.
-type SenderOption func(sender *Sender) error
+type (
+	// SenderOption specifies an option that can configure a Sender.
+	SenderOption func(sender *Sender) error
 
-// Sender is used to send messages as well as schedule them to be delivered at a later date.
-type Sender struct {
-	config struct {
+	// Sender is used to send messages as well as schedule them to be delivered at a later date.
+	Sender struct {
 		queueOrTopic string
+		links        internal.AMQPLinks
 	}
 
-	mu           *sync.Mutex
-	legacySender internal.LegacySender
-
-	linkState *linkState
-
-	// for testing
-	nsCreateLegacySender func(ctx context.Context, entityPath string) (internal.LegacySender, error)
-}
+	// SendableMessage are sendable using Sender.SendMessage.
+	// Message, MessageBatch implement this interface.
+	SendableMessage interface {
```

Comment:
Do we need to export this interface?  As a general rule, given the brittle nature of interfaces, we should strive to avoid exporting them.

Reply:
This interface is used for input only (which I thought was okay).

Reply:
What types implement this?  It seems like the methods on the interface are really internal implementation details.

Reply:
Message and MessageBatch (and when it comes in - an external form of *amqp.Message).

The other option, which might be more appropriate, is to have separate send functions for each type instead.

So:

SendMessage
SendMessageBatch
SendAMQPMessage

Would separate functions be better?

Reply:
The pattern here is to leave the interface type as exported but to _not_ export the methods since these are intended to be implemented internally.  This means that customers can't implement this interface (we don't want them to) thus allows us to change/extend it in a non-breaking way.

Reply:
I think I've got it. Does this seem right then (it's what I've just pushed):

```go
SendableMessage interface {
  toAMQPMessage() *amqp.Message
  messageType() string
}
```

Reply:
Yep!